pkg/controllers/tagging/tagging_controller.go

```diff
@@ -0,0 +1,248 @@
+	resources []string
+}
+
+// NewTaggingController creates a NewTaggingController object
+func NewTaggingController(
+	nodeInformer coreinformers.NodeInformer,
+	kubeClient clientset.Interface,
+	cloud cloudprovider.Interface,
+	nodeMonitorPeriod time.Duration,
+	tags map[string]string,
+	resources []string) (*TaggingController, error) {
+
+	awsCloud, ok := cloud.(*awsv1.Cloud)
+	if !ok {
+		err := fmt.Errorf("tagging controller does not support %v provider", cloud.ProviderName())
+		return nil, err
+	}
+
+	tc := &TaggingController{
+		nodeInformer:      nodeInformer,
+		kubeClient:        kubeClient,
+		cloud:             awsCloud,
+		nodeMonitorPeriod: nodeMonitorPeriod,
+		tags:              tags,
+		resources:         resources,
+		workqueue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Tagging"),
+		nodesSynced:       nodeInformer.Informer().HasSynced,
+	}
+
+	// Use shared informer to listen to add/update/delete of nodes. Note that any nodes
+	// that exist before tagging controller starts will show up in the update method
+	tc.nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { tc.enqueueNode(obj, tc.tagNodesResources) },
+		UpdateFunc: func(oldObj, newObj interface{}) { tc.enqueueNode(newObj, tc.tagNodesResources) },
+		DeleteFunc: func(obj interface{}) { tc.enqueueNode(obj, tc.untagNodeResources) },
+	})
+
+	return tc, nil
+}
+
+// Run will start the controller to tag resources attached to the cluster
+// and untag resources detached from the cluster.
+func (tc *TaggingController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer tc.workqueue.ShutDown()
+
+	// Wait for the caches to be synced before starting workers
+	klog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, tc.nodesSynced); !ok {
+		klog.Errorf("failed to wait for caches to sync")
+		return
+	}
+
+	klog.Infof("Starting the tagging controller")
+	go wait.Until(tc.work, tc.nodeMonitorPeriod, stopCh)
+
+	<-stopCh
+}
+
+// work is a long-running function that continuously
+// call process() for each message on the workqueue
+func (tc *TaggingController) work() {
+	for tc.Process() {
+	}
+}
+
+// Process reads each message in the queue and performs either
+// tag or untag function on the Node object
+func (tc *TaggingController) Process() bool {
+	obj, shutdown := tc.workqueue.Get()
+	if shutdown {
+		return false
+	}
+
+	klog.Infof("Starting to process %v", obj)
+
+	err := func(obj interface{}) error {
+		defer tc.workqueue.Done(obj)
+
+		workItem, ok := obj.(*workItem)
+		if !ok {
+			tc.workqueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected workItem in workqueue but got %#v", obj))
+			return nil
+		}
+
+		err := workItem.action(workItem.node)
+		if err != nil {
+			// Put the item back on the workqueue to handle any transient errors.
+			tc.workqueue.AddRateLimited(workItem)
```

Comment:
Does this happen on non-retryable errors too?  Like if the tag is never going to succeed are we going to keep adding it to the workqueue?

Reply:
Yes it does. Currently, we do not handle specific errors from EC2 but simply retrying everything with the assumption that no non-retryable errors expected here. Do you have any recommendation for the specific errors that we should look out for that could happen?

Reply:
I think we need to return nil from `workItem.action` for any non-retryable errors... and I'm not sure what possibilities there are... incorrectly formatted tag?

Reply:
We should log the error but we don't want to keep spamming the tag API, if possible.

Reply:
okay, I can try looking into `https://docs.aws.amazon.com/AWSEC2/latest/APIReference/errors-overview.html#CommonErrors` and see what errors are non-retryable in this case.